### PR TITLE
Add 6.2.0 upgrade notes about check configs for non-existent namespaces

### LINKED ISSUE #2921: Upgrading notes for 6.2.0 should cover known issue with orphaned check configs
### Description

Lingering check configurations which reference a non-existent namespace ("orphaned check configs") will cause backend to halt during upgrade process.

### Category

This is an issue with:

- [ ] errata (error in doc content)
- [ ] new doc (requires a completely new doc)
- [x] update (add missing or refresh existing content)
- [ ] improvement (improve docs content or format)

### Affected pages

https://docs.sensu.io/sensu-go/latest/operations/maintain-sensu/upgrade/

### Expected behavior

Known issues with upgrading are documented in upgrade notes.

### Current behavior

Upgrade blocker described in https://github.com/sensu/sensu-enterprise-go/issues/1432 is not documented in upgrade notes

### Corrections needed

Describe the conditions which could cause an error
Provide a one-liner or other simple way to determine if the conditions are present in a given installation 

### Additional context

Prior to Sensu Go 6.0, sensu-backend would allow a namespace to be deleted even when check configs and other resources still reference that namespace. 

Changes in Sensu Go 6.0 prevent namespaces from being deleted when other resources reference the namespace to be deleted. 

Changes in Sensu Go 6.2.0 require the backend to upgrade of check configurations, which upgrade operation fails when encountering a check that refers to a non-existent namespace.

Workaround: revert to 6.1.3, recreate the missing namespace(s), upgrade to 6.2.0 again.

### PR DESCRIPTION
**Pull request by hillaryfraley**
## Description
Adds a section about the reasons and solutions for 6.2.0 upgrade failure in the case of check configurations that reference non-existent namespaces.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2921

## Review Instructions
The issue mentions adding a one-liner to help users "...determine if the conditions are present in a given installation" but I wasn't quite sure what that might look like. I don't think there's a sensuctl command/subcommand that can do this job. Maybe it is something like dumping all checks and dumping all namespaces and then comparing?

### CODE REVIEW

**Comment by cwjohnston** [comment]
I did some testing and came up with a couple of curl + jq commands to identify existing namespaces and find orphaned checks via the API. Perhaps some of the content I've added here also belongs in troubleshooting? LMKWYT.

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 1]
```diff
@@ -11,6 +11,54 @@ menu:
...
 ---
 
+## Upgrade to Sensu Go 6.2.0 from any previous version
+
+As of Sensu Go 6.0, it is not possible to delete namespaces that were referenced in other resources.
+However, before Sensu Go 6.0, sensu-backend did allow you to delete a namespace even when other resources still referenced that namespace.
+As a result, pre-6.0 users may have resources, including check configurations, that reference non-existent namespaces.
+
+Upgrading to Sensu Go 6.2.0 requires sensu-backend to upgrade check configurations.
+If you have check configurations that reference non-existent namespaces, the 6.2.0 upgrade operation will fail with an error like this one when it encounters one of these check configurations:
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
If you have check configurations that reference non-existent namespaces, the 6.2.0 upgrade operation will fail when it encounters one of these check configurations.
You will see an error message like this:
```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 2]
```diff
@@ -11,6 +11,54 @@ menu:
...
+
+{{< code shell >}}
+{"component":"store-providers","error":"the namespace test does not exist","level":"error","msg":"error enabling round robin scheduling,backend restart required","time":"2020-12-27T08:41:59Z"}
+{{< /code >}}
+
+At this point, the backend will be effectively halted and subsequent restarts will result in the same state.
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
When this happens, the backend is effectively halted and subsequent restarts will result in the same state.
```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 3]
```diff
@@ -162,3 +210,4 @@ Then restart the backend.
 [4]: https://sensu.io/contact?subject=contact-sales/
 [5]: https://sensu.io/blog/one-year-of-sensu-go
 [6]: ../../../commercial/
+[7]: https://stedolan.github.io/jq/
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
[7]: https://stedolan.github.io/jq/
[8]: ../../../api/#authenticate-with-an-api-key

```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 4]
```diff
@@ -11,6 +11,54 @@ menu:
...
+Upgrading to Sensu Go 6.2.0 requires sensu-backend to upgrade check configurations.
+If you have check configurations that reference non-existent namespaces, the 6.2.0 upgrade operation will fail with an error like this one when it encounters one of these check configurations:
+
+{{< code shell >}}
+{"component":"store-providers","error":"the namespace test does not exist","level":"error","msg":"error enabling round robin scheduling,backend restart required","time":"2020-12-27T08:41:59Z"}
+{{< /code >}}
+
+At this point, the backend will be effectively halted and subsequent restarts will result in the same state.
+
+Assuming a Sensu backend running on localhost with the [jq utility][7] installed and an API key available as $SENSU_API_KEY, the following illustrates how to identify checks referencing deleted namespaces:
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md:29]:
```suggestion
The following commands use the [jq utility][7] to identify checks that reference deleted namespaces.
These commands use a Sensu backend running on `localhost` in the example URL and the environment variable $SENSU_API_KEY to represent a valid [API key][8].
```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 5]
```diff
@@ -11,6 +11,54 @@ menu:
...
+$ sensuctl namespace create test
+# Delete the check
+$ sensuctl check delete check-cpu --namespace test
+# Delete the now empty namespace
+$ sensuctl namespace delete test
+# You are now ready to upgrade to 6.2.0
+{{< /code >}}
+
+If you have already upgraded to 6.2.0, you can work around this issue by temporarily reverting your Sensu instance to Sensu Go 6.1.3.
+Then, recreate the missing namespaces referenced in your check configurations and upgrade again to 6.2.0.
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
### Remove checks that reference non-existent namespaces

Use the following commands with the [jq utility][7] to identify and remove checks that reference deleted namespaces before you upgrade to Sensu Go 6.2.0.

{{% notice note %}}
**NOTE**: If you have already upgraded to Sensu Go 6.2.0, you can work around this issue by temporarily reverting your Sensu instance to Sensu Go 6.1.3.
Then, recreate the missing namespaces referenced in your check configurations and upgrade again to 6.2.0.
{{% /notice %}}

These commands use a Sensu backend running on `localhost` in the example URL and the environment variable $SENSU_API_KEY to represent a valid [API key][8].

1. Get a list of existing namespaces.
In this example, the existing namespaces are `stage` and `dev`.

   {{< code shell >}}
$ curl -s -H "Authorization: Key $SENSU_API_KEY" http://localhost:8080/api/core/v2/namespaces | jq '[.[].name]'
[
   "stage",
   "dev"
]
{{< /code >}}

2. Print the name and namespace for any checks that reference a namespace that is not specified in the jq expression on the same line.
Your jq expression should include all of the namespaces you retrieved in step 1 (in this example, `["stage","dev"]`).

   {{< code shell >}}
$ curl -s -H "Authorization: Key $SENSU_API_KEY" http://localhost:8080/api/core/v2/checks | jq '["stage","dev"] as $valid | .[] | select(.metadata.namespace as $in | $valid | index($in) | not) | {name: .metadata.name, namespace: .metadata.namespace}'
{
  "name": "check-cpu",
  "namespace": "test"
}
{{< /code >}}

3. Recreate the missing `test` namespace so you can delete the `check-cpu` check.

   {{< code shell >}}
$ sensuctl namespace create test
{{< /code >}}

4. Delete the `check-cpu` check.

   {{< code shell >}}
$ sensuctl check delete check-cpu --namespace test
{{< /code >}}

5. Delete the `test` namespace, which is now empty after you deleted `check-cpu` in step 4.

   {{< code shell >}}
$ sensuctl namespace delete test
{{< /code >}}

After completing these commands, you can upgrade to 6.2.0.
```

Comment (hillaryfraley) [reply content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]: This diff is terrible, I'm sorry. I basically made this a subsection, reformatted the commands into separate numbered steps, and moved the workaround to a note at the top of the subsection.

**Comment by hillaryfraley** [comment]
@cwjohnston thank you! This is awesome. I made some formatting changes because I thought it made the commands easier to follow -- wdyt?

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 6]
```diff
@@ -11,6 +11,54 @@ menu:
     parent: maintain-sensu
 ---
 
+## Upgrade to Sensu Go 6.2.0 from any previous version
+
+As of Sensu Go 6.0, it is not possible to delete namespaces that were referenced in other resources.
+However, before Sensu Go 6.0, sensu-backend did allow you to delete a namespace even when other resources still referenced that namespace.
+As a result, pre-6.0 users may have resources, including check configurations, that reference non-existent namespaces.
```
Comment (cwjohnston) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
Prior to Sensu Go 6.0, sensu-backend allowed you to delete a namespace even when other resources still referenced that namespace. 

As of Sensu Go 6.0, it is not possible to delete namespaces that are referenced in other resources.

As a result, users whose configuration predates Sensu Go 6.0 may have lingering resources, including check configurations, that reference non-existent namespaces.
```

**Review by cwjohnston — CHANGES_REQUESTED** [review]

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 7]
```diff
@@ -11,6 +11,77 @@ menu:
     parent: maintain-sensu
 ---
 
+## Upgrade to Sensu Go 6.2.0 from any previous version
+
+Prior to Sensu Go 6.0, sensu-backend allowed you to delete a namespace even when other resources still referenced that namespace. 
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 8]
```diff
@@ -11,6 +11,77 @@ menu:
...
+As of Sensu Go 6.0, it is not possible to delete namespaces that are referenced in other resources.
+
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
```

File: content/sensu-go/6.2/operations/maintain-sensu/upgrade.md [hunk 9]
```diff
@@ -11,6 +11,75 @@ menu:
...
+{{< /code >}}
+
+When this happens, the backend is effectively halted and subsequent restarts will result in the same state.
+
+### Remove checks that reference non-existent namespaces
+
+Use the following commands with the [jq utility][7] to identify and remove checks that reference deleted namespaces before you upgrade to Sensu Go 6.2.0.
+
+{{% notice note %}}
+**NOTE**: If you have already upgraded to Sensu Go 6.2.0, you can work around this issue by temporarily reverting your Sensu instance to Sensu Go 6.1.3.
```
Comment (hillaryfraley) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]: Does it have to be 6.1.3? Or would it be 6.1.4 (the last release before 6.2.0) instead?

Comment (cwjohnston) [reply content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]: I believe 6.1.4 would work.

Comment (cwjohnston) [inline content/sensu-go/6.2/operations/maintain-sensu/upgrade.md]:
```suggestion
**NOTE**: If you have already upgraded to Sensu Go 6.2.0, you can work around this issue by temporarily reverting your Sensu instance to Sensu Go 6.1.4.
```

**Review by cwjohnston — APPROVED** [review]

**Pull request merged**